src/services/rgbpp.ts

```diff
@@ -298,9 +306,87 @@ export default class RgbppCollector extends BaseQueueWorker<IRgbppCollectRequest
     return null;
   }
 
+  async isIsomorphicTx(
+    btcTx: Transaction,
+    ckbTx: CKBComponents.RawTransaction,
+    validateCommitment?: boolean,
+  ): Promise<boolean> {
+    // Find the commitment from the btc_tx
+    const btcTxCommitment = tryGetCommitmentFromBtcTx(btcTx);
+    if (!btcTxCommitment) {
+      return false;
+    }
+
+    // Check inputs:
+    // 1. Find the last index of the type inputs
+    // 2. Check if all rgbpp_lock inputs can be found in the btc_tx.vin
+    // 3. Check if the inputs contain at least one rgbpp_lock cell (as L1-L1 and L1-L2 transactions should have)
+    const inputs = await this.cradle.ckb.getInputCellsByOutPoint(ckbTx.inputs.map((input) => input.previousOutput!));
+    const lastTypeInputIndex = findLastIndex(inputs, (input) => !!input.cellOutput.type);
+    const anyRgbppLockInput = inputs.some((input) => isRgbppLock(input.cellOutput.lock));
+    if (!anyRgbppLockInput) {
+      return false;
+    }
+    const allInputsValid = inputs.every((input, index) => {
+      if (!input.cellOutput.type) {
+        return true;
+      }
+      if (!isRgbppLock(input.cellOutput.lock)) {
+        return true;
+      }
+      const btcInput = btcTx.vin[index];
+      const rgbppLockArgs = unpackRgbppLockArgs(input.cellOutput.lock.args);
+      return btcInput && btcInput.txid === remove0x(rgbppLockArgs.btcTxid) && btcInput.vout === rgbppLockArgs.outIndex;
+    });
+    if (!allInputsValid) {
+      return false;
+    }
+
+    // Check outputs:
+    // 1. Find the last index of the type outputs, and check if at least one type output exists
+    // 2. Check if all type outputs are rgbpp_lock or btc_time_lock cells
+    // 4. Check if each rgbpp_lock cell has an isomorphic UTXO in the btc_tx.vout
```

Comment:
`check if at least one type output exists` may not be necessary.

See https://github.com/ckb-cell/rgbpp/blob/4abd216dbd5d25b8f59e843de39643f9b7513f9d/contracts/rgbpp-lock/src/main.rs#L81-L84

Reply:
So it seems the outputs would be valid even if it includes no rgbpp-lock or btc-time-lock cells, am I right?

Reply:
Fixed at e0c1c51

Reply:
> So it seems the outputs would be valid even if it includes no rgbpp-lock or btc-time-lock cells, am I right?

Yes.
IMO, if no output has type_script, then rgbpp-lock or btc-time-lock is not required.